src/jit/simdcodegenxarch.cpp

```diff
@@ -840,13 +1004,548 @@ void CodeGen::genSIMDIntrinsicUnOp(GenTreeSIMD* simdNode)
+    }
+    genProduceReg(simdNode);
+}
+
+//--------------------------------------------------------------------------------
+// genSIMDIntrinsicNarrow: Generate code for SIMD Intrinsic Narrow operations
+//
+// Arguments:
+//    simdNode - The GT_SIMD node
+//
+// Notes:
+//    This intrinsic takes two arguments. The first operand is narrowed to produce the
+//    lower elements of the results, and the second operand produces the high elements.
+//
+void CodeGen::genSIMDIntrinsicNarrow(GenTreeSIMD* simdNode)
+{
+    assert(simdNode->gtSIMDIntrinsicID == SIMDIntrinsicNarrow);
+
+    GenTree*  op1       = simdNode->gtGetOp1();
+    GenTree*  op2       = simdNode->gtGetOp2();
+    var_types baseType  = simdNode->gtSIMDBaseType;
+    regNumber targetReg = simdNode->gtRegNum;
+    assert(targetReg != REG_NA);
+    var_types      simdType = simdNode->TypeGet();
+    emitAttr       emitSize = emitTypeSize(simdType);
+    InstructionSet iset     = compiler->getSIMDInstructionSet();
+
+    genConsumeOperands(simdNode);
+    regNumber op1Reg = op1->gtRegNum;
+    regNumber op2Reg = op2->gtRegNum;
+    if (baseType == TYP_DOUBLE)
+    {
+        regNumber tmpReg;
+        if (iset == InstructionSet_AVX)
+        {
+            regMaskTP tmpRegsMask = simdNode->gtRsvdRegs & RBM_ALLFLOAT;
+            regMaskTP tmpReg1Mask = genFindLowestBit(tmpRegsMask);
+            tmpReg                = genRegNumFromMask(tmpReg1Mask);
+        }
+        else
+        {
+            tmpReg = genRegNumFromMask(simdNode->gtRsvdRegs & RBM_ALLFLOAT);
+        }
+        inst_RV_RV(INS_cvtpd2ps, targetReg, op1Reg, simdType);
+        inst_RV_RV(INS_cvtpd2ps, tmpReg, op2Reg, simdType);
+        // Now insert the high-order result (in tmpReg) into the upper half of targetReg.
+        if (compiler->canUseAVX())
+        {
+            getEmitter()->emitIns_R_R_I(INS_vinsertf128, EA_32BYTE, targetReg, tmpReg, 0x01);
+        }
+        else
+        {
+            inst_RV_RV_IV(INS_shufps, EA_16BYTE, targetReg, tmpReg, SHUFFLE_YXYX);
+        }
+    }
+    else if (varTypeIsLong(baseType))
+    {
+        if (iset == InstructionSet_AVX)
+        {
+            // We have 8 long elements, 0-3 in op1Reg, 4-7 in op2Reg.
+            // We will generate the following:
+            //   vextracti128 tmpReg, op1Reg, 1       (extract elements 2 and 3 into tmpReg)
+            //   vextracti128 tmpReg2, op2Reg, 1      (extract elements 6 and 7 into tmpReg2)
+            //   vinserti128  tmpReg, tmpReg2, 1       (insert elements 6 and 7 into the high half of tmpReg)
+            //   mov          tmpReg2, op1Reg
+            //   vinserti128  tmpReg2, op2Reg, 1      (insert elements 4 and 5 into the high half of tmpReg2)
+            //   pshufd       tmpReg, tmpReg, XXZX    ( -  - 7L 6L  -  - 3L 2L) in tmpReg
+            //   pshufd       tgtReg, tmpReg2, XXZX   ( -  - 5L 4L  -  - 1L 0L) in tgtReg
+            //   punpcklqdq   tgtReg, tmpReg
```

Comment:
These kind of high-level descriptions would be useful for understanding the logic.